src/rdkafka_msgset_writer.c

```diff
@@ -0,0 +1,1160 @@
+                rkm->rkm_key ? rkm->rkm_key_len :
+                (size_t)RD_KAFKAP_BYTES_LEN_NULL);
+        sz_ValueLen = rd_uvarint_enc_i32(
+                varint_ValueLen, sizeof(varint_ValueLen),
+                rkm->rkm_payload ? rkm->rkm_len :
+                (size_t)RD_KAFKAP_BYTES_LEN_NULL);
+        sz_HeaderCount = rd_uvarint_enc_i32(
+                varint_HeaderCount, sizeof(varint_HeaderCount), 0);
+
+        /* Calculate MessageSize without length of Length (added later)
+         * to store it in Length. */
+        MessageSize =
+                1 /* MsgAttributes */ +
+                sz_TimestampDelta +
+                sz_OffsetDelta +
+                sz_KeyLen +
+                rkm->rkm_key_len +
+                sz_ValueLen +
+                rkm->rkm_len +
+                sz_HeaderCount;
+
+        /* Length */
+        sz_Length = rd_uvarint_enc_i64(varint_Length, sizeof(varint_Length),
+                                       MessageSize);
+        rd_kafka_buf_write(rkbuf, varint_Length, sz_Length);
+        MessageSize += sz_Length;
+
+        /* Attributes */
+        rd_kafka_buf_write_i8(rkbuf, MsgAttributes);
+
+        /* TimestampDelta */
+        rd_kafka_buf_write(rkbuf, varint_TimestampDelta, sz_TimestampDelta);
+
+        /* OffsetDelta */
+        rd_kafka_buf_write(rkbuf, varint_OffsetDelta, sz_OffsetDelta);
+
+        /* KeyLen */
+        rd_kafka_buf_write(rkbuf, varint_KeyLen, sz_KeyLen);
+
+        /* Key (if any) */
+        if (rkm->rkm_key)
+                rd_kafka_buf_write(rkbuf, rkm->rkm_key, rkm->rkm_key_len);
+
+        /* ValueLen */
+        rd_kafka_buf_write(rkbuf, varint_ValueLen, sz_ValueLen);
+
+        /* Write or copy Value/payload */
+        if (rkm->rkm_payload)
+                rd_kafka_msgset_writer_write_msg_payload(msetw, rkm, free_cb);
+
+        /* HeaderCount (headers currently not implemented) */
+        rd_kafka_buf_write(rkbuf, varint_HeaderCount, sz_HeaderCount);
+
+        /* Return written message size */
+        return MessageSize;
+}
+
+
+/**
+ * @brief Write message to messageset buffer.
+ * @returns the number of bytes written.
+ */
+static size_t
+rd_kafka_msgset_writer_write_msg (rd_kafka_msgset_writer_t *msetw,
+                                  rd_kafka_msg_t *rkm,
+                                  int64_t Offset, int8_t MsgAttributes,
+                                  void (*free_cb)(void *)) {
+        size_t outlen;
+        size_t (*writer[]) (rd_kafka_msgset_writer_t *,
+                            rd_kafka_msg_t *, int64_t, int8_t,
+                            void (*)(void *)) = {
+                [0] = rd_kafka_msgset_writer_write_msg_v0_1,
+                [1] = rd_kafka_msgset_writer_write_msg_v0_1,
+                [2] = rd_kafka_msgset_writer_write_msg_v2
+        };
+        size_t actual_written;
+        size_t pre_pos;
+
+        if (likely(rkm->rkm_timestamp))
+                MsgAttributes |= RD_KAFKA_MSG_ATTR_CREATE_TIME;
```

Comment:
Just to be clear, v2 does not have a timestamp type attribute at the message level; it is only supported in the message set attributes.